### ISSUE TITLE
Any update on shard imbalance issue for parquet dataset?

### ISSUE THREAD

**Issue by PHILO-HE** [uber/petastorm#719]
According to the discussions in github, it looks that petastorm shard is row-group based for parquet data. i.e., the smallest division unit is row-group. Considering row-groups may not be evenly assigned and each row-group size is not equal generally, therefore, in distributed training scenario, there may exist evident imbalance in terms of dataset shard for training processes, which may cause issues in distributed training. I noted, there is already a detailed discussion in last year in #508. I am writing to see whether petastorm currently has any new functionality to fix or workaround this imbalance issue. 
Thanks in advance!

**Comment by selitvin** [comment]
Thanks for bringing this topic up. There are no new features. I wonder if option 3 as discussed in #508 does not work for your case? If not, then why?

**Comment by PHILO-HE** [comment]
Thanks so much for your response!
I am not sure whether the suggested option 3 is good to me. I am using petastorm's APIs for pytorch. I note an option named `shuffle_row_groups` can control whether to shuffle row groups. And I note another option named `shuffling_queue_capacity` for `DataLoader` can be used to uniformly sample the fetched data kept in a queue. I guess `shuffle_row_groups` should be turned on and `shuffling_queue_capacity` should be set a proper value to achieve shuffle in petastorm. Right?

One concern from me is converge performance. The shuffle approach may not get similar converge speed as ideal shard in distributed training. If dataset is not properly shuffled, it may lead to a kind of bias for training, considering certain number of samples can probably be feed into two or more trainers in one epoch.

**Comment by selitvin** [comment]
That could be true. One data point: from working with several different teams that compared the approach on some of our models, no difference was observed. Of course this does not have to hold for other models.

**Comment by PHILO-HE** [comment]
I prefer to load a large batch of samples, and assign a mini batch of samples to each rank according to rank ID. In this way, extra useless data is loaded for a rank. In my tests, I found `BatchedDataLoader` can accelerate data loading, especially for large batch size. So no much overhead was observed.
The stated shuffle approach may be good to me also. I will have a try to compare the performance.

**Comment by PHILO-HE** [comment]
An implementation to my last comment. In my mentioned approach, each worker will fetch a global size of batch, but just feed a shard of batches (1/_n_ of global size, if there is _n_ worker) into its training model. This approach requires each worker to get same global size of batch, also same ordered, for each iteration. We found that petastorm data loader worker number must be 1 to meet the stated requirement. 
Even though extra useless data is loaded for each worker, it was observed that the overhead is tolerable in the case of using `BatchedDataLoader`.

**Comment by PHILO-HE** [comment]
I am closing this issue. Please re-open it if further discussion is needed.
